contents/javascript/javascript.md

```diff
@@ -0,0 +1,272 @@
+
+As a beginner, there may be some hesistation to use lodash or external utility libraries. The big advantage is the amount of benefit it brings to the team. Whether its the overhead of having to load the `lodash` library, or having to learn a large library's syntax, not having to maintain your own utility functions makes the project more manageable, maintainable and that alone makes it worth its cost.
+
+## Testing Libraries
+
+There are a large amount of Javascript testing libraries. The below is a non-exhaustive list.
+
+- [Mocha](https://mochajs.org/)
+- [Jasmine](https://jasmine.github.io/)
+- [QUnitJs](https://qunitjs.com)
+- [Jest](https://facebook.github.io/jest/)
+- [AVA](https://github.com/avajs/ava)
+- [Tape](https://github.com/substack/tape)
+
+So which one do you choose? Among the above, I have had the opportunity of write in all of the above except Tape. 
+
+My general advice is to go for Mocha if you are new, due to its [large community](http://stateofjs.com/2016/testing/) and therefore overall amount of help available. When you are familiar with Mocha and understand its pitfalls (speed and organisation), you can explore other libraries such as Jest.
+
+Jest, due to its speed, support for React.js, support for asynchronous testing and helpful terminal outputs, seems to be the upcoming major player. The fact that it is being made by Facebook is icing to the cake.
+
+The odd test framework in the list above is QUnit, which is a test framework that runs on the browser instead of node.js like the others. QUnit has been around longer. Unless you are running JQuery and need to test for browser or UI related bugs like in [TEAMMATES](https://github.com/TEAMMATES/teammates/), it is no longer a conventional choice.
+
+## Resources
+[Clean Code Javascript](https://github.com/ryanmcdermott/clean-code-javascript)
+Apparently most of what I wrote appears in this huge guide in some form. It's an amazing resource and also explains SOLID clearly near the bottom.
+
+[Airbnb Javascript Style](https://github.com/airbnb/javascript)
+It is one thing to follow the style guide, and another to understand why it is that way. Understanding why Airbnb chose certain constructs and syntax reveals ways to write code that is clean, understandable and maintainable.
+
+[JavaScript Design Patterns](https://addyosmani.com/resources/essentialjsdesignpatterns/book/)
+The title says it all. From the most common to obscure patterns, this book covers design patterns and explains trade offs. Although it specifically caters to Javascript, it's recommended reading for all prospective software engineers.
+
+## References
+[Namespacing in Javascript](https://javascriptweblog.wordpress.com/2010/12/07/namespacing-in-javascript/)
+
+[MMethod Chaining in JavaScript](https://schier.co/blog/2013/11/14/method-chaining-in-javascript.html)
```

Comment:
Technically, _references_ should be referenced explicitly in the main text. For example, something like this:

> The tool X is currently the most popular tool in the marked [source: [X Dev Guide]()]. However, tool Y is also ...
> ...
> ### References
> * [X Dev Guide] [_A Guide to Developing on X_]() - This is the official Developer Guide for X